src/tyr/isochrone_serializer.cc

```diff
@@ -12,117 +12,193 @@ using namespace valhalla::baldr::json;
 
 namespace {
 using rgba_t = std::tuple<float, float, float>;
+
+using namespace valhalla;
+using namespace tyr;
+using namespace midgard;
+using contour_t = std::list<PointLL>;                 // single ring
+using feature_t = std::list<contour_t>;               // rings per interval
+using contours_t = std::vector<std::list<feature_t>>; // all rings
+using contour_group_t = std::vector<const contour_t*>;
+using grouped_contours_t = std::vector<contour_group_t>;
+using contour_interval_t = std::tuple<size_t, float, std::string, std::string>;
+
+grouped_contours_t GroupContours(const bool polygons, const feature_t& contours) {
+  grouped_contours_t results;
+
+  // if the user requested linestrings, we'll give them linestrings
+  if (!polygons || contours.size() < 2) {
+    std::for_each(contours.begin(), contours.end(),
+                  [&results](const contour_t& c) { results.push_back({&c}); });
+    return results;
+  }
+
+  // inner rings have negative area
+  auto isExteriorRing = [](const contour_t& c) -> bool { return polygon_area(c) > 0; };
+
+  std::vector<const contour_t*> inner_ptrs;
+  std::for_each(contours.begin(), contours.end(),
+                [&results, &inner_ptrs, isExteriorRing](const contour_t& c) {
+                  if (isExteriorRing(c)) {
+                    results.push_back({&c});
+                  } else {
+                    inner_ptrs.push_back(&c);
+                  }
+                });
+
+  // exactly one exterior ring, so all inners go in one group
+  if (results.size() == 1) {
+    std::for_each(inner_ptrs.begin(), inner_ptrs.end(),
+                  [&results](const contour_t* c) { results[0].push_back(c); });
+    return results;
+  }
+
+  // iterate over outer rings and for each inner ring check if the inner ring is within the exterior
+  // ring
+  for (const auto inner : inner_ptrs) {
+
+    // get the first point of the ring (could be any though)
+    PointLL inner_pt = inner->front();
+    bool found_exterior = false;
+    // go over exterior rings from smallest to largest
+    for (size_t i = results.size(); i > 0; --i) {
+      contour_t ext = *results[i - 1][0];
```

Comment:
if you don't want the pointer, rather take a ref to the underlying value or you do a copy
```suggestion
      const contour_t& ext = *results[i - 1][0];
```